### DISCUSSION TITLE
Get compiled CSS path or contents for injection?

### DISCUSSION THREAD

**Discussion by nnx0r** [wxt-dev/wxt#542] [Q&A] [answered]
Is it possible to get either the path to the CSS file of content script or the file contents? 
I need to somehow get it so that I can inject it into shadow-root of an Web Component on a webpage.

Not sure how to achieve this.

**Comment by nnx0r** [comment]
Apparently, this won't work:
`browser.runtime.getURL('content-scripts/inject.css')`

Injecting in shadow-root: 
- `<link ref="stylesheet" href="chrome-extension://blhphmohaaddbepnpajbbijjfhpgppco/content-scripts/inject.css" />` - does not affect elements in shadow-root.
- `<style>...</style>` - using `fetch` to fetch text contents from the chrome-extension css path also doesn't work

In both cases, I get an error:        
GET chrome-extension://invalid/

Not sure why, because path is correct...

**Reply by aklinker1** [reply]
`createShadowRootUi` do this automatically for you. There's a couple of setup stepps, so checkout the docs:

https://wxt.dev/guide/content-script-ui.html#shadow-root

Otherwise, if you want to manually setup a shadow root, you can set `cssInjectionMode` to `"manual"`, and the css file should be available at `browser.runtime.getURL("/content-scripts/<name>.css")`.

```ts
// entrypoints/injected/index.ts
import "./style.css"; // The CSS to generate alongside the entrypoint

export default defineContentScript({
  cssInjectionMode: "manual",
  async main() {
    const res = await fetch(browser.runtime.getURL("/content-scripts/injected.css"));
    const css = await res.text();
    
    // Use CSS in <style> tag body.
  },
});
```

**Reply by nnx0r** [reply]
Hmmm, seems like `createShadowRootUi` is for content-script itself, so that styles are contained within shadow-root for the extension's UI. 

What I'm looking for is inject extension style inside of Web Component that could exist on any web page (e.g. Disney Plus)

I want to inject extension CSS globally and then inject that same CSS inside of `shadow-root` of an Web Component.

I am not sure why this throws an error for me:
```
const res = await fetch(browser.runtime.getURL("/content-scripts/injected.css"));
const css = await res.text();
```

**Reply by aklinker1** [reply]
> I am not sure why this throws an error for me:
> ```
> const res = await fetch(browser.runtime.getURL("/content-scripts/injected.css"));
> const css = await res.text();
> ```

Oh, you probably need to add it to `web_accesible_rseources`

**Reply by aklinker1** [reply]
Use `cssInjectionMode: "ui"` will do that for you automatically

**Reply by nnx0r** [reply]
With `cssInjectionMode: 'ui'` I no longer have CSS that globally affects web pages, which I require.
However, with it, the fetch of `content-scripts/inject.css` works fine, so it must be what you told me one answer earlier: `web_accesible_resources` should be declared.

Do I just add that in my wxt config (manifest)?

**Reply by nnx0r** [reply]
Looks about right... Fetches the resource now that I can inject into shadowed-root.

```js
{
  web_accessible_resources: [{
    resources: ['/content-scripts/inject.css'],
    matches: hosts,
  }],
}
```

What will happen in Firefox MV2?

**Comment by nnx0r** [comment, marked as answer]
Delightful... Seems like these are the necessary steps to support MV2 & MV3. I stripped away all of the other configuration that I have, that isn't relevant to the question.

```ts
export default defineConfig({
  manifest: ({ manifestVersion }) => {
    const hosts = ['https://*/*', 'http://*/*']

    const config: UserManifest = {
      web_accessible_resources: [
        {
          resources: ['/content-scripts/inject.css'],
          matches: hosts,
        },
      ],
    }

    if (manifestVersion === 2) {
      config.web_accessible_resources = ['/content-scripts/inject.css']
    }

    return config
  },
})
```

And then in `inject.content.ts` _(or whatever the filename is)_

```ts
export default defineContentScript({
  allFrames: true,
  matches: ['https://*/*', 'http://*/*'],
  async main() {
    // ...
    
    let inlineCSS = await fetch(
      browser.runtime.getURL('/content-scripts/inject.css'),
    ).then((r) => r.text())
  }
})
```

Thanks again @aklinker1 , you rock! 🤘

**Reply by aklinker1** [reply]
Glad you got it working!

FYI, you can just provide `web_accessible_resources` in the MV3 format, and WXT will convert them to a `string[]` for MV2 automatically.

```ts
export default defineConfig({
  manifest: {
    web_accessible_resources: [
      {
        resources: ['/content-scripts/inject.css'],
        matches: ['https://*/*', 'http://*/*'],
      },
    ],
  },
})
```

Don't remember which version I added that feature in.

**Reply by nnx0r** [reply]
I think I tried that, before my MV2 override. Didn't work! Firefox complained that it wants `string[]`. 

Either I missed something or I'm not on the version that got this update. Will update to most recent tomorrow!

**Reply by nnx0r** [reply]
Just a quick update. It works... Updated to latest, but went through the releases first to make sure that breaking changes are resolved on my end.

You released that update on this version:
[v0.16.4](https://github.com/wxt-dev/wxt/releases/tag/v0.16.4)

Thanks again!
